ArjunaCore/arjuna/tests/byteman-scripts/reaper.btm

```diff
@@ -329,3 +329,49 @@ DO debug("removing latches on reaper and reaper worker"),
    deleteRendezvous("reaper1", 2),
    deleteRendezvous("reaperworker1", 2)   
 ENDRULE
+
+#########################################################################
+# ReaperTestCase4 syncs the shutdown with something being removed
+#
+
+
+RULE ReaperTestCase4 rendezvous in shutdown
+CLASS com.arjuna.ats.arjuna.coordinator.TransactionReaper
+METHOD shutdown
+AT CALL wait
+BIND NOTHING
+IF TRUE
+DO rendezvous("controlledshutdown")
+ENDRULE
+
+RULE Set a flag
+CLASS com.arjuna.ats.arjuna.coordinator.TransactionReaper
+METHOD insert
+AT ENTRY
+BIND NOTHING
+IF TRUE
+DO flag("shutdown")
+ENDRULE
+
+RULE Return false (only once)
+CLASS com.arjuna.ats.arjuna.coordinator.TransactionReaper
+METHOD removeElementClient
+AT ENTRY
+BIND NOTHING
+IF flagged("shutdown") && $0._inShutdown
+DO rendezvous("controlledshutdown"),
+   debug("**** BYTEMAN **** -> removeElementClient AT ENTRY: value of _inShutdown is " + $0._inShutdown),
+   $0._inShutdown = false,
```

Comment:
@jmfinelli IIUC this is simulating the assumption that _inShutdown reads false out of a cache due to being non-volatile

Reply:
Yep, that's the reason of this byteman rule. I'll make it more explicit

Reply:
Actually, having run it in the debugger, I can find that another thread has found that it is _inShutdown and is waking up the waiting terminate call: https://github.com/jmfinelli/narayana/blob/7ee80804e9bd13364d57d22bc67d3499f74336cf/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L828

Reply:
Interesting. That makes the second reapable element in `ReaperTestCase4.test()` useless. So we only need to defend ourselves from the situation where we have only one reapable, and it gets removed using `removeElementClient`

Reply:
> Actually, having run it in the debugger, I can find that another thread has found that it is _inShutdown and is waking up the waiting terminate call: https://github.com/jmfinelli/narayana/blob/7ee80804e9bd13364d57d22bc67d3499f74336cf/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java#L828

This makes totally sense and demonstrates that the TransactionReaper does exactly what it is supposed to do